In [1]:
import pandas as pd
import numpy as np

In [2]:
dataset_names = ['dim_campaigns', 'dim_delivery_locations', 'dim_places', 'dim_tables', 'dim_taxonomy_terms', 'dim_users', 'fct_bonus_codes', 'fct_campaigns', 'fct_order_items', 'fct_orders', 'most_ordered']
datasets = {}
for name in dataset_names:
    datasets[name] = pd.read_csv(name + '.csv', low_memory=False)

In [36]:
datasets['dim_places'].columns[datasets['dim_places'].isna().all()]

Index(['kitchen_staff_ids', 'nets_mid', 'realm', 'delivery_qr_code_link',
       'takeaway_qr_code_link', 'theme'],
      dtype='object')

In [3]:
# Overview of all datasets
print("=" * 60)
print("DATASET OVERVIEW FOR SHIFT SCHEDULER APPLICATION")
print("=" * 60)

for name, df in datasets.items():
    print(f"\n{'='*60}")
    print(f"📊 {name.upper()}")
    print(f"{'='*60}")
    print(f"Shape: {df.shape[0]:,} rows × {df.shape[1]} columns")
    print(f"\nColumns: {list(df.columns)}")
    print(f"\nData Types:\n{df.dtypes.value_counts()}")
    print(f"\nFirst 3 rows:")
    display(df.head(3))

DATASET OVERVIEW FOR SHIFT SCHEDULER APPLICATION

📊 DIM_CAMPAIGNS
Shape: 641 rows × 7 columns

Columns: ['id', 'user_id', 'created', 'updated', 'place_id', 'status', 'type']

Data Types:
int64    5
str      2
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated  place_id    status  \
0  60594      423  1612819150  1659482246     59897  Inactive   
1  60609      423  1612819657  1659482250     59897  Inactive   
2  60610      423  1612820393  1659482250     59821  Inactive   

                              type  
0  Discount on specific menu items  
1  Discount on specific menu items  
2   2 for 1 on specific menu items


📊 DIM_DELIVERY_LOCATIONS
Shape: 3,906 rows × 12 columns

Columns: ['id', 'user_id', 'created', 'updated', 'address', 'area_id', 'country', 'latitude', 'longitude', 'status', 'tag', 'deleted']

Data Types:
int64      5
str        4
float64    3
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated                 address  area_id  \
0  60210      476  1612377196  1612377196     Næsborgvej 10, 1 TV  11450.0   
1  60250      476  1612379452  1612379452        Tuborgvej 24 1.4  11473.0   
2  60944      646  1613318361  1615125053  A F Beyersvej 23, 2.th  11457.0   

  country  latitude  longitude status    tag  deleted  
0     NaN       NaN        NaN    NaN   Home        0  
1     NaN       NaN        NaN    NaN  Other        0  
2     NaN       NaN        NaN    NaN   Home        0


📊 DIM_PLACES
Shape: 1,824 rows × 189 columns

Columns: ['id', 'user_id', 'created', 'updated', 'title', 'activated', 'active', 'bankrupt', 'binding_period', 'chain_id', 'contact_email', 'contact_mobile_phone', 'contract_start', 'country', 'currency', 'cvr_number', 'duplicate', 'ecommerce_fee', 'eu_commission', 'invoicing_start_date', 'isv_commission', 'isv_partner', 'non_eu_commission', 'onboarded_by', 'payment_terminal_provider', 'area_id', 'processing_fee', 'sales_outcome_id', 'sales_stage', 'service_charge', 'setup_type', 'termination_date', 'termination_value', 'transferred_contract', 'type_id', 'cashier_ids', 'manager_ids', 'driver_ids', 'payment_terminal_ids', 'kitchen_staff_ids', 'contact_name', 'description', 'cuisine_ids', 'trainee_mode', 'demo_mode', 'default_order_type_cashier', 'default_order_type_customer', 'show_bestsellers', 'use_quick_search', 'display_cashier_images', 'display_created_by', 'use_customer_facing_display', 'timezone', 'contact_language', 'business_name',

id  user_id     created     updated       title  activated  active  \
0  59813      417  1610718144  1744221291  Green Love          0       0   
1  59817      417  1610718304  1744221291    Grønnere          0       0   
2  59821      417  1610866418  1744221291   Eden Jaxx          0       0   

   bankrupt  binding_period  chain_id  ...   vat  vip_threshold  \
0         0              36       NaN  ...  25.0             20   
1         0              36       NaN  ...  25.0             20   
2         0              36       NaN  ...  25.0             20   

   visit_duration votes waiting_time unspecified_section_id  \
0              90    22           25                5972947   
1              90     0           15                5972948   
2              90     8           25                5972949   

   unspecified_item_id  enable_sms_on_order  sms_on_order_text  \
0               306775                    0                NaN   
1               306777                    0                NaN   
2               306779                    0                NaN   

   exclude_from_ops_report  
0                        0  
1                        0  
2                        0  

[3 rows x 189 columns]


📊 DIM_TABLES
Shape: 3,400 rows × 18 columns

Columns: ['id', 'user_id', 'created', 'updated', 'title', 'bookable', 'capacity', 'category_id', 'external_id', 'link', 'place_id', 'status', 'table_stand', 'qr_code', 'deleted', 'floorplan', 'shape', 'table_section_id']

Data Types:
int64      8
str        8
float64    2
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated title  bookable  capacity  \
0  81734      417  1646230164  1731431107     1         1         4   
1  81735      417  1646230184  1731431107     2         1         4   
2  81736      417  1646230200  1731431108     3         1         4   

   category_id external_id                                               link  \
0      12185.0         NaN  http://lovingloyalty.com/order/table-1/zula-ve...   
1      12185.0         NaN  http://lovingloyalty.com/order/table-2/zula-ve...   
2      12185.0         NaN  http://lovingloyalty.com/order/table-3/zula-ve...   

   place_id  status                         table_stand  \
0     64276  Active  public://tables/64276/81734_en.pdf   
1     64276  Active  public://tables/64276/81735_en.pdf   
2     64276  Active  public://tables/64276/81736_en.pdf   

                           qr_code  deleted                         floorplan  \
0  public://tables/64276/81734.png        0  {"top":720,"left":100,"angle":0}   
1  public://tables/64276/81735.png        0  {"top":160,"left":180,"angle":0}   
2  public://tables/64276/81736.png        0  {"top":440,"left":180,"angle":0}   

       shape  table_section_id  
0  rectangle               NaN  
1  rectangle               NaN  
2  rectangle               NaN


📊 DIM_TAXONOMY_TERMS
Shape: 904 rows × 6 columns

Columns: ['id', 'user_id', 'created', 'updated', 'vocabulary', 'name']

Data Types:
int64    4
str      2
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated vocabulary    name
0     1     1351  1746612556  1746612556     gender    Male
1     2     1351  1746612556  1746612556     gender  Female
2  1332     1351  1746612556  1746612556       type     Bar


📊 DIM_USERS
Shape: 22,955 rows × 41 columns

Columns: ['id', 'user_id', 'created', 'updated', 'account_closure_requested', 'age_group_id', 'app_version', 'area_id', 'cltv', 'country', 'currency', 'date_of_birth', 'do_not_contact', 'first_name', 'gender_id', 'last_name', 'mobile_phone', 'picture', 'orders', 'redeemed_points', 'referring_user_id', 'referring_place_id', 'savings', 'source', 'type', 'api_key', 'roles', 'email', 'password', 'referral_id', 'otp', 'language', 'email_temp', 'mobile_phone_temp', 'mobile_phone_valid', 'email_valid', 'barcode_scanner_ids', 'receipt_printer_ids', 'payment_terminal_ids', 'notifications', 'favorite_place_ids']

Data Types:
str        18
float64    13
int64      10
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated  account_closure_requested  \
0    0      0.0  1571676937  1605685627                          0   
1    1      1.0  1571676937  1744002851                          0   
2  417    417.0  1610542930  1752828166                          0   

   age_group_id  app_version  area_id   cltv country  ... language  \
0           NaN          NaN      NaN    0.0      DK  ...       en   
1       12033.0          NaN  11470.0  232.0      DK  ...       en   
2           NaN          NaN      NaN  993.0      DK  ...       en   

   email_temp  mobile_phone_temp mobile_phone_valid  email_valid  \
0         NaN                NaN                  0            0   
1         NaN                NaN                  0            0   
2         NaN                NaN                  1            0   

  barcode_scanner_ids  receipt_printer_ids payment_terminal_ids  \
0                 NaN                  NaN                  NaN   
1                 NaN                  NaN                  NaN   
2                 NaN       443164|6009243              6009244   

                                       notifications  favorite_place_ids  
0                                                NaN                 NaN  
1  push_orders|push_others|email_orders|email_others                 NaN  
2  push_reviews|push_deals|push_orders|push_other...                 NaN  

[3 rows x 41 columns]


📊 FCT_BONUS_CODES
Shape: 6 rows × 9 columns

Columns: ['id', 'user_id', 'created', 'updated', 'end_date_time', 'start_date_time', 'place_id', 'points', 'redemptions']

Data Types:
int64      7
float64    2
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated  end_date_time  start_date_time  \
0  61628      423  1614960382  1615136541     1615248000       1614902400   
1  61922      423  1615559087  1637324088     1640822400       1615420800   
2  61940      423  1615564762  1622661882     1640908800       1615420800   

   place_id  points  redemptions  
0       NaN     150          NaN  
1   59897.0     200          NaN  
2       NaN     300          NaN


📊 FCT_CAMPAIGNS
Shape: 641 rows × 55 columns

Columns: ['id', 'user_id', 'created', 'updated', 'title', 'age_group_ids', 'auto_created', 'item_ids', 'code', 'account_ids', 'delivery', 'eat_in', 'takeaway', 'discount', 'discount_type', 'start_date_time', 'end_date_time', 'freebie_id', 'monday_from', 'monday_to', 'tuesday_from', 'tuesday_to', 'wednesday_from', 'wednesday_to', 'thursday_from', 'thursday_to', 'friday_from', 'friday_to', 'saturday_from', 'saturday_to', 'sunday_from', 'sunday_to', 'gender_ids', 'goodie_bag_id', 'maximum_cltv', 'maximum_orders', 'minimum_cltv', 'minimum_orders', 'minimum_spend', 'parent_id', 'area_ids', 'place_id', 'price', 'provider', 'redemptions', 'redemptions_per_customer', 'status', 'table_id', 'tier_category_ids', 'type', 'used_redemptions', 'validity_hours', 'value', 'variation', 'visibility']

Data Types:
float64    37
int64       9
str         9
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated                               title  \
0  60594      423  1612819150  1659482246        15% discount on Snickerskage   
1  60609      423  1612819657  1659482250              10% discount on Burger   
2  60610      423  1612820393  1659482250  2 for 1 on Organic Homemade Shakes   

   age_group_ids  auto_created item_ids code  account_ids  ...  \
0            NaN           NaN    59932  NaN          NaN  ...   
1            NaN           NaN    60454  NaN          NaN  ...   
2            NaN           NaN    59830  NaN          NaN  ...   

   redemptions_per_customer    status  table_id  tier_category_ids  \
0                       NaN  Inactive       NaN                NaN   
1                       NaN  Inactive       NaN                NaN   
2                       NaN  Inactive       NaN                NaN   

                              type  used_redemptions  validity_hours  value  \
0  Discount on specific menu items                 0             NaN    NaN   
1  Discount on specific menu items                 0             NaN    NaN   
2   2 for 1 on specific menu items                 0             NaN    NaN   

   variation  visibility  
0        NaN         NaN  
1        NaN         NaN  
2        NaN         NaN  

[3 rows x 55 columns]


📊 FCT_ORDER_ITEMS
Shape: 1,999,341 rows × 23 columns

Columns: ['id', 'user_id', 'created', 'updated', 'title', 'campaign_id', 'commission_amount', 'cost', 'discount_amount', 'external_id', 'group', 'instructions', 'item_id', 'order_id', 'points_earned', 'points_redeemed', 'price', 'quantity', 'redemptions', 'removed_ingredients', 'add_on_ids', 'status', 'vat_amount']

Data Types:
float64    10
int64       7
str         6
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated             title  campaign_id  \
0  60824      512  1613139429  1708921636             Latte          NaN   
1  60837      514  1613144783  1708921895  The Classic Meal          NaN   
2  60838      514  1613144783  1708921895  The Classic Meal          NaN   

   commission_amount   cost  discount_amount external_id  ...  order_id  \
0               1.50   50.0              0.0         NaN  ...   60825.0   
1               5.25  175.0              0.0         NaN  ...   60841.0   
2               4.80  160.0              0.0         NaN  ...   60841.0   

  points_earned  points_redeemed  price  quantity  redemptions  \
0           NaN              NaN   25.0         2            0   
1           NaN              NaN  175.0         1            0   
2           NaN              NaN  160.0         1            0   

   removed_ingredients  add_on_ids   status vat_amount  
0                  NaN         NaN  Pending        NaN  
1                  NaN       59995  Pending        NaN  
2                  NaN         NaN  Pending        NaN  

[3 rows x 23 columns]


📊 FCT_ORDERS
Shape: 399,810 rows × 40 columns

Columns: ['id', 'user_id', 'created', 'updated', 'updated_by', 'account_id', 'cash_amount', 'cashier_notified', 'channel', 'customer_mobile_phone', 'customer_name', 'code', 'delivery_charge', 'delivery_location_id', 'demo_mode', 'discount_amount', 'driver_id', 'external_id', 'instructions', 'items_amount', 'payment_method', 'pickup_time', 'place_id', 'points_earned', 'points_redeemed', 'promise_time', 'rejection_reason', 'referring_user_id', 'service_charge', 'source', 'split_bill', 'split_bill_type', 'status', 'synchronized_to_accounting', 'table_id', 'tier_id', 'total_amount', 'trainee_mode', 'type', 'vat_amount']

Data Types:
float64    25
str        11
int64       4
Name: count, dtype: int64

First 3 rows:


id  user_id     created     updated  updated_by  account_id  \
0  60825      512  1613139429  1687470923         NaN         NaN   
1  60841      514  1613144783  1687470923         NaN         NaN   
2  60852      489  1613149610  1687470923         NaN         NaN   

   cash_amount  cashier_notified channel  customer_mobile_phone  ...  \
0         50.0               NaN     App                    NaN  ...   
1        655.0               NaN     App                    NaN  ...   
2        430.0               NaN     App                    NaN  ...   

  split_bill split_bill_type  status  synchronized_to_accounting  table_id  \
0        NaN             NaN  Closed                         NaN       NaN   
1        NaN             NaN  Closed                         NaN       NaN   
2        NaN             NaN  Closed                         NaN       NaN   

   tier_id  total_amount trainee_mode      type  vat_amount  
0  59898.0          50.0          NaN  Takeaway       12.50  
1  59822.0         655.0          NaN  Takeaway      163.75  
2  59822.0         430.0          NaN  Takeaway      107.50  

[3 rows x 40 columns]


📊 MOST_ORDERED
Shape: 95,435 rows × 7 columns

Columns: ['place_id', 'store_address', 'device_serial', 'device_product', 'item_id', 'item_name', 'order_count']

Data Types:
int64      3
float64    2
str        2
Name: count, dtype: int64

First 3 rows:


place_id  store_address device_serial  device_product  item_id  \
0     59813            NaN           NaN             NaN    60067   
1     59813            NaN           NaN             NaN    67188   
2     59813            NaN           NaN             NaN    67190   

           item_name  order_count  
0           Knækbrød           10  
1       Salt mandler            7  
2  Marinerede oliven            7

In [4]:
# Key columns analysis for Shift Scheduler Application
print("=" * 70)
print("🗓️  KEY COLUMNS FOR SHIFT SCHEDULER APPLICATION")
print("=" * 70)

# Analyze fct_orders for time patterns
orders = datasets['fct_orders']
orders['created_dt'] = pd.to_datetime(orders['created'], unit='s')
orders['hour'] = orders['created_dt'].dt.hour
orders['day_of_week'] = orders['created_dt'].dt.day_name()

print("\n📌 1. FCT_ORDERS - CRITICAL FOR DEMAND FORECASTING")
print("-" * 50)
print(f"   Total orders: {len(orders):,}")
print(f"   Date range: {orders['created_dt'].min()} to {orders['created_dt'].max()}")
print(f"\n   🔑 Key Columns:")
print("   • created       → Order timestamp (demand patterns)")
print("   • place_id      → Location identifier")
print("   • type          → Order type (Eat-in/Takeaway/Delivery)")
print("   • status        → Order status")
print("   • table_id      → Table assignment")
print("   • total_amount  → Revenue (for prioritizing busy periods)")

print(f"\n   Order Types Distribution:")
print(orders['type'].value_counts().to_string())

print(f"\n   Orders by Day of Week:")
print(orders['day_of_week'].value_counts().to_string())

print(f"\n   Peak Hours (orders count):")
print(orders['hour'].value_counts().sort_index().head(10).to_string())

# Analyze dim_places
places = datasets['dim_places']
print("\n\n📌 2. DIM_PLACES - LOCATION/VENUE INFO")
print("-" * 50)
print(f"   Total places: {len(places):,}")
print(f"\n   🔑 Key Columns:")
print("   • id, title           → Place identification")
print("   • type_id             → Type of venue (Restaurant/Bar/Café)")
print("   • opening_hours_*     → Operating hours (Mon-Sun)")
print("   • address, latitude, longitude → Location")
print("   • status, active      → Operational status")

# Check opening hours columns
opening_cols = [c for c in places.columns if 'opening' in c.lower() or 'hour' in c.lower()]
print(f"\n   Opening Hours Columns: {opening_cols[:10]}")

# Analyze dim_users for staff
users = datasets['dim_users']
print("\n\n📌 3. DIM_USERS - STAFF/EMPLOYEE DATA")
print("-" * 50)
print(f"   Total users: {len(users):,}")
print(f"\n   🔑 Key Columns:")
print("   • id, first_name, last_name → Employee identification")
print("   • roles                     → Job roles (cashier, manager, etc.)")
print("   • mobile_phone, email       → Contact info")
print("   • area_id                   → Work area/location")

print(f"\n   Sample Roles (from 'roles' column):")
roles = users['roles'].dropna().str.split('|').explode().value_counts()
print(roles.head(10).to_string())

# Analyze dim_tables
tables = datasets['dim_tables']
print("\n\n📌 4. DIM_TABLES - TABLE/SEATING CAPACITY")
print("-" * 50)
print(f"   Total tables: {len(tables):,}")
print(f"\n   🔑 Key Columns:")
print("   • id, title          → Table identification")
print("   • capacity           → Seating capacity")
print("   • place_id           → Which venue")
print("   • status             → Active/Inactive")
print("   • bookable           → Can be reserved")

print(f"\n   Table Capacity Distribution:")
print(tables['capacity'].value_counts().sort_index().to_string())

# Analyze campaigns for special events
campaigns = datasets['fct_campaigns']
print("\n\n📌 5. FCT_CAMPAIGNS - PROMOTIONS & EVENTS")
print("-" * 50)
print(f"   Total campaigns: {len(campaigns):,}")
print(f"\n   🔑 Key Columns:")
print("   • start_date_time, end_date_time → Campaign period")
print("   • place_id                       → Affected venue")
print("   • monday_from to sunday_to       → Day-specific timing")
print("   • type                           → Campaign type")
print("   (Campaigns may require extra staff)")

🗓️  KEY COLUMNS FOR SHIFT SCHEDULER APPLICATION

📌 1. FCT_ORDERS - CRITICAL FOR DEMAND FORECASTING
--------------------------------------------------
   Total orders: 399,810
   Date range: 2021-02-12 14:17:09 to 2024-02-16 19:14:57

   🔑 Key Columns:
   • created       → Order timestamp (demand patterns)
   • place_id      → Location identifier
   • type          → Order type (Eat-in/Takeaway/Delivery)
   • status        → Order status
   • table_id      → Table assignment
   • total_amount  → Revenue (for prioritizing busy periods)

   Order Types Distribution:
type
Takeaway    347397
Eat In       51964
Delivery       428

   Orders by Day of Week:
day_of_week
Friday       79374
Saturday     65712
Thursday     60751
Wednesday    56781
Tuesday      50626
Monday       44571
Sunday       41995

   Peak Hours (orders count):
hour
0     3265
1     1998
2     1060
3      326
4       56
5      430
6     3048
7     7649
8    11712
9    16861


📌 2. DIM_PLACES - LOCATION/VENUE INFO
----------

In [5]:
# Summary: Most Important Columns for Shift Scheduler
print("=" * 70)
print("📋 SUMMARY: PRIORITY COLUMNS FOR SHIFT SCHEDULING")
print("=" * 70)

summary_data = {
    'Dataset': ['fct_orders', 'fct_orders', 'fct_orders', 'fct_orders', 
                'dim_places', 'dim_places', 'dim_places',
                'dim_users', 'dim_users', 'dim_users',
                'dim_tables', 'dim_tables',
                'fct_campaigns', 'fct_campaigns'],
    'Column': ['created', 'place_id', 'type', 'total_amount',
               'id/title', 'opening_hours_*', 'type_id',
               'id/first_name/last_name', 'roles', 'area_id',
               'capacity', 'place_id',
               'start/end_date_time', 'day_from/day_to'],
    'Purpose': ['Demand forecasting by hour/day', 'Link orders to locations', 
                'Dine-in vs delivery staffing', 'Revenue-based priority',
                'Identify venues', 'Operating hours for shifts', 'Venue type affects staffing',
                'Employee info', 'Assign roles (cashier/manager)', 'Employee work location',
                'Calculate covers/capacity', 'Link tables to venue',
                'Plan for promotions', 'Day-specific campaign hours']
}

summary_df = pd.DataFrame(summary_data)
print("\n")
display(summary_df)

# Key metrics for shift scheduling
print("\n" + "=" * 70)
print("📊 KEY METRICS FOR SHIFT SCHEDULING")
print("=" * 70)

print(f"\n📍 Venues (dim_places): {len(datasets['dim_places']):,}")
print(f"👥 Users (dim_users): {len(datasets['dim_users']):,}")
print(f"🪑 Tables (dim_tables): {len(datasets['dim_tables']):,}")
print(f"📝 Orders (fct_orders): {len(datasets['fct_orders']):,}")
print(f"🎯 Campaigns (fct_campaigns): {len(datasets['fct_campaigns']):,}")

# Staff with operational roles
staff_roles = ['cashier', 'manager', 'pos_device', 'agent']
operational_staff = datasets['dim_users'][
    datasets['dim_users']['roles'].fillna('').str.contains('|'.join(staff_roles), case=False)
]
print(f"\n👷 Operational Staff (cashier/manager/pos): {len(operational_staff):,}")

📋 SUMMARY: PRIORITY COLUMNS FOR SHIFT SCHEDULING




Dataset                   Column                         Purpose
0      fct_orders                  created  Demand forecasting by hour/day
1      fct_orders                 place_id        Link orders to locations
2      fct_orders                     type    Dine-in vs delivery staffing
3      fct_orders             total_amount          Revenue-based priority
4      dim_places                 id/title                 Identify venues
5      dim_places          opening_hours_*      Operating hours for shifts
6      dim_places                  type_id     Venue type affects staffing
7       dim_users  id/first_name/last_name                   Employee info
8       dim_users                    roles  Assign roles (cashier/manager)
9       dim_users                  area_id          Employee work location
10     dim_tables                 capacity       Calculate covers/capacity
11     dim_tables                 place_id            Link tables to venue
12  fct_campaigns      start/end_date_time             Plan for promotions
13  fct_campaigns          day_from/day_to     Day-specific campaign hours


📊 KEY METRICS FOR SHIFT SCHEDULING

📍 Venues (dim_places): 1,824
👥 Users (dim_users): 22,955
🪑 Tables (dim_tables): 3,400
📝 Orders (fct_orders): 399,810
🎯 Campaigns (fct_campaigns): 641

👷 Operational Staff (cashier/manager/pos): 1,894


In [6]:
# Business Nature Analysis
print("=" * 70)
print("🏢 BUSINESS NATURE ANALYSIS")
print("=" * 70)

# 1. Venue Types
places = datasets['dim_places']
taxonomy = datasets['dim_taxonomy_terms']

# Get venue types from taxonomy
venue_types = taxonomy[taxonomy['vocabulary'] == 'type'][['id', 'name']]
print("\n📍 VENUE TYPES IN THE PLATFORM:")
print("-" * 50)
print(venue_types['name'].values)

# 2. Cuisines offered
cuisines = taxonomy[taxonomy['vocabulary'] == 'cuisine']['name'].values
print(f"\n🍽️ CUISINES OFFERED: {len(cuisines)} types")
print("-" * 50)
print(f"Sample: {list(cuisines[:15])}...")

# 3. Order channels and types
orders = datasets['fct_orders']
print("\n📱 ORDER CHANNELS:")
print("-" * 50)
print(orders['channel'].value_counts().to_string())

print("\n🛍️ ORDER TYPES:")
print("-" * 50)
print(orders['type'].value_counts().to_string())

# 4. Geographic scope
print("\n🌍 GEOGRAPHIC SCOPE:")
print("-" * 50)
areas = taxonomy[taxonomy['vocabulary'] == 'area']['name'].values
print(f"Areas served: {len(areas)}")
print(f"Sample areas: {list(areas[:10])}")

# Check country from delivery locations
delivery = datasets['dim_delivery_locations']
print(f"\nCountries: {delivery['country'].dropna().unique()}")

# 5. Campaign/Promotion Types
print("\n🎯 PROMOTION TYPES:")
print("-" * 50)
print(datasets['fct_campaigns']['type'].value_counts().to_string())

# 6. Customer Tiers (Loyalty Program)
tiers = taxonomy[taxonomy['vocabulary'] == 'tier_category']['name'].values
print(f"\n⭐ LOYALTY TIERS: {list(tiers)}")

# 7. Business Model Indicators
print("\n💼 BUSINESS MODEL INDICATORS:")
print("-" * 50)
print(f"• Places with tables: {places['id'].nunique():,} venues")
print(f"• Total table capacity: {datasets['dim_tables']['capacity'].sum():,} seats")
print(f"• Users with loyalty points: {(datasets['dim_users']['redeemed_points'] > 0).sum():,}")
print(f"• Campaigns/Promotions run: {len(datasets['fct_campaigns']):,}")
print(f"• Bonus codes issued: {len(datasets['fct_bonus_codes']):,}")

# Revenue analysis
print(f"\n💰 REVENUE INDICATORS:")
print(f"• Total orders: {len(orders):,}")
print(f"• Average order value: {orders['total_amount'].mean():.2f} DKK")
print(f"• Total revenue: {orders['total_amount'].sum():,.0f} DKK")

# 8. Staff roles
print("\n👥 STAFF ROLES IN SYSTEM:")
print("-" * 50)
all_roles = datasets['dim_users']['roles'].dropna().str.split('|').explode().value_counts()
print(all_roles.to_string())

🏢 BUSINESS NATURE ANALYSIS

📍 VENUE TYPES IN THE PLATFORM:
--------------------------------------------------
<StringArray>
['Bar', 'Café', 'Lounge', 'Restaurant', 'Pub', 'Unknown']
Length: 6, dtype: str

🍽️ CUISINES OFFERED: 148 types
--------------------------------------------------
Sample: ['Albanian', 'Andhra', 'Anglo-Indian', 'Middle Eastern', 'Argentine', 'Armenian', 'Assyrian', 'Awadhi', 'Azerbaijani', 'Balochi', 'Bangladeshi', 'Belarusian', 'Bengali', 'Berber', 'Brazilian']...

📱 ORDER CHANNELS:
--------------------------------------------------
channel
App    358321
Web     41489

🛍️ ORDER TYPES:
--------------------------------------------------
type
Takeaway    347397
Eat In       51964
Delivery       428

🌍 GEOGRAPHIC SCOPE:
--------------------------------------------------
Areas served: 706
Sample areas: ['1264 København K', '1620 København V', '1635 København V', '2000 Frederiksberg', '2100 København Ø', '2200 København N', '2300 København S', '2500 Valby', '2720 Vanløs

## 🏢 Business Summary

### What is this business?
This is a **Restaurant/Hospitality Technology Platform** operating in **Denmark (Copenhagen area)** that provides:

1. **Multi-venue Management** - Supports 1,800+ venues including:
   - Restaurants, Bars, Cafés, Lounges, Pubs

2. **Order Management System** - Handles:
   - Dine-in (Eat-in) orders with table management
   - Takeaway orders
   - Delivery orders with location tracking

3. **Customer Loyalty Program** - Features:
   - Points-based rewards system
   - Tiered membership (Gold, Silver, Bronze)
   - Bonus codes and promotional campaigns

4. **POS & Staff Management** - Includes:
   - Cashier, Manager, and POS device roles
   - Table assignments and capacity tracking

### Key Business Characteristics
| Aspect | Details |
|--------|---------|
| **Industry** | Food & Beverage / Hospitality Tech |
| **Geography** | Denmark (Copenhagen metro area) |
| **Currency** | DKK (Danish Krone) |
| **Business Model** | B2B2C Platform (connects venues with customers) |
| **Channels** | Mobile App, POS, Web |
| **Cuisines** | 100+ types (Danish, Italian, Asian, etc.) |

### Platform Name
Based on URLs in the data: **"Loving Loyalty"** (lovingloyalty.com) - A loyalty and ordering platform for restaurants.

# 🧙 Shift Wizard Solution Design
## QuickServe Kitchens: Intelligent Shift Planning

Based on the available data, here's how a solution should address the core challenges.

In [7]:
# =============================================================================
# SHIFT WIZARD SOLUTION DESIGN - Data Mapping Analysis
# =============================================================================

import matplotlib.pyplot as plt
from datetime import datetime

print("=" * 80)
print("🧙 SHIFT WIZARD: DATA-TO-SOLUTION MAPPING")
print("=" * 80)

# =============================================================================
# 1. DEMAND PREDICTION ENGINE - What data supports this?
# =============================================================================
print("\n" + "=" * 80)
print("📊 1. DEMAND PREDICTION ENGINE")
print("=" * 80)

orders = datasets['fct_orders'].copy()
orders['created_dt'] = pd.to_datetime(orders['created'], unit='s')
orders['date'] = orders['created_dt'].dt.date
orders['hour'] = orders['created_dt'].dt.hour
orders['day_of_week'] = orders['created_dt'].dt.dayofweek
orders['day_name'] = orders['created_dt'].dt.day_name()
orders['month'] = orders['created_dt'].dt.month
orders['week'] = orders['created_dt'].dt.isocalendar().week

print("\n✅ AVAILABLE DATA FOR DEMAND FORECASTING:")
print("-" * 50)
print(f"   • Historical orders: {len(orders):,} records")
print(f"   • Date range: {orders['created_dt'].min().date()} to {orders['created_dt'].max().date()}")
print(f"   • Unique venues: {orders['place_id'].nunique():,}")

# Hourly demand pattern
hourly_demand = orders.groupby('hour').size()
print(f"\n   📈 Peak hours identified: {hourly_demand.idxmax()}:00 ({hourly_demand.max():,} orders)")
print(f"   📉 Slowest hour: {hourly_demand.idxmin()}:00 ({hourly_demand.min():,} orders)")

# Day of week pattern
daily_demand = orders.groupby('day_name').size().reindex(
    ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
)
print(f"\n   📅 Busiest day: {daily_demand.idxmax()} ({daily_demand.max():,} orders)")
print(f"   📅 Slowest day: {daily_demand.idxmin()} ({daily_demand.min():,} orders)")

# Order type breakdown (affects staffing mix)
print(f"\n   🍽️ Order Type Mix (staffing implications):")
order_type_pct = (orders['type'].value_counts(normalize=True) * 100).round(1)
for otype, pct in order_type_pct.items():
    print(f"      • {otype}: {pct}%")

print("\n⚠️ DATA GAPS FOR DEMAND PREDICTION:")
print("-" * 50)
print("   • No social media/viral event tracking")
print("   • No weather data correlation")
print("   • No local event calendar integration")
print("   • No foot traffic sensors")

🧙 SHIFT WIZARD: DATA-TO-SOLUTION MAPPING

📊 1. DEMAND PREDICTION ENGINE

✅ AVAILABLE DATA FOR DEMAND FORECASTING:
--------------------------------------------------
   • Historical orders: 399,810 records
   • Date range: 2021-02-12 to 2024-02-16
   • Unique venues: 359

   📈 Peak hours identified: 16:00 (45,452 orders)
   📉 Slowest hour: 4:00 (56 orders)

   📅 Busiest day: Friday (79,374 orders)
   📅 Slowest day: Sunday (41,995 orders)

   🍽️ Order Type Mix (staffing implications):
      • Takeaway: 86.9%
      • Eat In: 13.0%
      • Delivery: 0.1%

⚠️ DATA GAPS FOR DEMAND PREDICTION:
--------------------------------------------------
   • No social media/viral event tracking
   • No weather data correlation
   • No local event calendar integration
   • No foot traffic sensors


In [8]:
# =============================================================================
# 2. OPTIMAL STAFFING CALCULATOR - What data supports this?
# =============================================================================
print("\n" + "=" * 80)
print("👥 2. OPTIMAL STAFFING CALCULATOR")
print("=" * 80)

users = datasets['dim_users']
tables = datasets['dim_tables']

# Staff analysis
print("\n✅ AVAILABLE STAFF DATA:")
print("-" * 50)

# Extract staff by role
staff_with_roles = users[users['roles'].notna()].copy()
all_roles_list = staff_with_roles['roles'].str.split('|').explode()
role_counts = all_roles_list.value_counts()

print(f"   Total users in system: {len(users):,}")
print(f"\n   Staff by Role:")
for role, count in role_counts.items():
    if role in ['cashier', 'manager', 'pos_device', 'agent', 'editor']:
        print(f"      • {role}: {count:,}")

# Operational staff (relevant for scheduling)
operational_roles = ['cashier', 'manager', 'pos_device']
operational_mask = users['roles'].fillna('').str.contains('|'.join(operational_roles), case=False)
operational_staff = users[operational_mask]
print(f"\n   🎯 Schedulable Staff (cashier/manager/pos): {len(operational_staff):,}")

# Capacity analysis
print(f"\n✅ VENUE CAPACITY DATA:")
print("-" * 50)
print(f"   Total tables: {len(tables):,}")
print(f"   Total seating capacity: {tables['capacity'].sum():,} seats")
print(f"   Avg capacity per table: {tables['capacity'].mean():.1f} seats")
print(f"   Unique venues with tables: {tables['place_id'].nunique():,}")

# Calculate staffing ratios from historical data
print(f"\n📐 DERIVED STAFFING METRICS:")
print("-" * 50)
# Orders per staff member (rough indicator)
avg_daily_orders = len(orders) / orders['date'].nunique()
print(f"   Avg daily orders (all venues): {avg_daily_orders:,.0f}")
print(f"   Avg orders per venue per day: {avg_daily_orders / orders['place_id'].nunique():.1f}")

# Peak hour analysis for staffing
peak_orders_per_venue = orders[orders['hour'] == 16].groupby('place_id').size().mean()
print(f"   Avg orders per venue at peak (16:00): {peak_orders_per_venue:.1f}")

print("\n⚠️ DATA GAPS FOR STAFFING OPTIMIZATION:")
print("-" * 50)
print("   • No staff availability/preferences data")
print("   • No shift history or actual schedules")
print("   • No labor cost data (hourly rates)")
print("   • No service time metrics (orders/staff/hour)")
print("   • No skills/certifications per employee")


👥 2. OPTIMAL STAFFING CALCULATOR

✅ AVAILABLE STAFF DATA:
--------------------------------------------------
   Total users in system: 22,955

   Staff by Role:
      • cashier: 1,771
      • pos_device: 437
      • manager: 63
      • editor: 22
      • agent: 10

   🎯 Schedulable Staff (cashier/manager/pos): 1,887

✅ VENUE CAPACITY DATA:
--------------------------------------------------
   Total tables: 3,400
   Total seating capacity: 12,350 seats
   Avg capacity per table: 3.6 seats
   Unique venues with tables: 222

📐 DERIVED STAFFING METRICS:
--------------------------------------------------
   Avg daily orders (all venues): 400
   Avg orders per venue per day: 1.1
   Avg orders per venue at peak (16:00): 154.1

⚠️ DATA GAPS FOR STAFFING OPTIMIZATION:
--------------------------------------------------
   • No staff availability/preferences data
   • No shift history or actual schedules
   • No labor cost data (hourly rates)
   • No service time metrics (orders/staff/hour)
   •

In [9]:
# =============================================================================
# 3. CALL-OFF / RAPID ADJUSTMENT MODULE
# =============================================================================
print("\n" + "=" * 80)
print("🚨 3. CALL-OFF & RAPID ADJUSTMENT MODULE")
print("=" * 80)

print("\n✅ WHAT WE CAN BUILD WITH CURRENT DATA:")
print("-" * 50)
print("   • Staff contact directory (mobile_phone, email)")
print("   • Role-based replacement matching (cashier→cashier)")
print("   • Area-based proximity matching (area_id)")

# Staff contact availability
staff_with_contact = users[
    (users['mobile_phone'].notna()) | (users['email'].notna())
]
print(f"\n   Staff with contact info: {len(staff_with_contact):,}")

# Role distribution for replacement pool
print(f"\n   Replacement Pool by Role:")
for role in ['cashier', 'manager', 'pos_device']:
    count = users['roles'].fillna('').str.contains(role, case=False).sum()
    print(f"      • {role}: {count:,} potential replacements")

print("\n⚠️ DATA GAPS FOR CALL-OFF MANAGEMENT:")
print("-" * 50)
print("   • No PTO/leave request history")
print("   • No call-off history patterns")
print("   • No employee reliability scores")
print("   • No real-time availability tracking")
print("   • No notification preferences")

# =============================================================================
# 4. CAMPAIGN/EVENT SURGE PLANNING
# =============================================================================
print("\n" + "=" * 80)
print("🎯 4. CAMPAIGN/EVENT SURGE PLANNING")
print("=" * 80)

campaigns = datasets['fct_campaigns']
campaigns['start_dt'] = pd.to_datetime(campaigns['start_date_time'], unit='s', errors='coerce')
campaigns['end_dt'] = pd.to_datetime(campaigns['end_date_time'], unit='s', errors='coerce')

print("\n✅ CAMPAIGN DATA FOR SURGE PREDICTION:")
print("-" * 50)
print(f"   Total campaigns: {len(campaigns):,}")
print(f"   Campaign types:")
for ctype, count in campaigns['type'].value_counts().items():
    print(f"      • {ctype}: {count}")

# Day-specific scheduling columns
day_cols = [c for c in campaigns.columns if any(day in c.lower() for day in 
            ['monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday', 'sunday'])]
print(f"\n   Day-specific timing columns: {day_cols[:4]}...")

# Campaign impact on demand
print(f"\n📈 CAMPAIGN IMPACT ANALYSIS:")
campaigns_with_redemptions = campaigns[campaigns['used_redemptions'] > 0]
print(f"   Campaigns with redemptions: {len(campaigns_with_redemptions)}")
print(f"   Total redemptions: {campaigns['used_redemptions'].sum():.0f}")
print(f"   Avg redemptions per campaign: {campaigns_with_redemptions['used_redemptions'].mean():.1f}")

print("\n⚠️ DATA GAPS FOR EVENT PLANNING:")
print("-" * 50)
print("   • No external event calendar")
print("   • No social media monitoring")
print("   • No weather data integration")
print("   • No foot traffic correlation")


🚨 3. CALL-OFF & RAPID ADJUSTMENT MODULE

✅ WHAT WE CAN BUILD WITH CURRENT DATA:
--------------------------------------------------
   • Staff contact directory (mobile_phone, email)
   • Role-based replacement matching (cashier→cashier)
   • Area-based proximity matching (area_id)

   Staff with contact info: 22,955

   Replacement Pool by Role:
      • cashier: 1,771 potential replacements
      • manager: 63 potential replacements
      • pos_device: 437 potential replacements

⚠️ DATA GAPS FOR CALL-OFF MANAGEMENT:
--------------------------------------------------
   • No PTO/leave request history
   • No call-off history patterns
   • No employee reliability scores
   • No real-time availability tracking
   • No notification preferences

🎯 4. CAMPAIGN/EVENT SURGE PLANNING

✅ CAMPAIGN DATA FOR SURGE PREDICTION:
--------------------------------------------------
   Total campaigns: 641
   Campaign types:
      • 2 for 1 on specific menu items: 339
      • All you can eat: 189
      

In [10]:
# =============================================================================
# 5. LABOR COST vs SERVICE QUALITY BALANCER
# =============================================================================
print("\n" + "=" * 80)
print("💰 5. LABOR COST vs SERVICE QUALITY BALANCER")
print("=" * 80)

print("\n✅ REVENUE/VALUE DATA AVAILABLE:")
print("-" * 50)
print(f"   Total revenue in dataset: {orders['total_amount'].sum():,.0f} DKK")
print(f"   Average order value: {orders['total_amount'].mean():.2f} DKK")
print(f"   Median order value: {orders['total_amount'].median():.2f} DKK")

# Revenue by hour (to prioritize staffing)
hourly_revenue = orders.groupby('hour')['total_amount'].sum()
print(f"\n   💎 Highest revenue hour: {hourly_revenue.idxmax()}:00 ({hourly_revenue.max():,.0f} DKK)")
print(f"   💸 Lowest revenue hour: {hourly_revenue.idxmin()}:00 ({hourly_revenue.min():,.0f} DKK)")

# Revenue by day
daily_revenue = orders.groupby('day_name')['total_amount'].sum().reindex(
    ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
)
print(f"\n   📅 Revenue by Day:")
for day, rev in daily_revenue.items():
    print(f"      • {day}: {rev:,.0f} DKK")

print("\n⚠️ DATA GAPS FOR COST OPTIMIZATION:")
print("-" * 50)
print("   • No labor cost data (hourly wages)")
print("   • No staff productivity metrics")
print("   • No service time measurements")
print("   • No customer satisfaction scores")
print("   • No wait time tracking")

# =============================================================================
# 6. VENUE-SPECIFIC DEMAND PATTERNS
# =============================================================================
print("\n" + "=" * 80)
print("🏪 6. VENUE-SPECIFIC DEMAND PATTERNS")
print("=" * 80)

# Top venues by order volume
top_venues = orders.groupby('place_id').agg({
    'id': 'count',
    'total_amount': 'sum'
}).rename(columns={'id': 'order_count'}).sort_values('order_count', ascending=False).head(10)

places = datasets['dim_places']
top_venues = top_venues.merge(places[['id', 'title']], left_index=True, right_on='id', how='left')

print("\n📊 TOP 10 VENUES BY ORDER VOLUME:")
print("-" * 50)
for _, row in top_venues.iterrows():
    print(f"   {row['title'][:30]:30} | {row['order_count']:,} orders | {row['total_amount']:,.0f} DKK")

# Venue type affects staffing needs
print(f"\n🏢 VENUE TYPES (from taxonomy):")
venue_types = datasets['dim_taxonomy_terms'][datasets['dim_taxonomy_terms']['vocabulary'] == 'type']
for _, row in venue_types.iterrows():
    print(f"   • {row['name']}")


💰 5. LABOR COST vs SERVICE QUALITY BALANCER

✅ REVENUE/VALUE DATA AVAILABLE:
--------------------------------------------------
   Total revenue in dataset: 178,090,177 DKK
   Average order value: 445.44 DKK
   Median order value: 80.00 DKK

   💎 Highest revenue hour: 16:00 (102,322,272 DKK)
   💸 Lowest revenue hour: 4:00 (3,849 DKK)

   📅 Revenue by Day:
      • Monday: 5,327,433 DKK
      • Tuesday: 100,399,867 DKK
      • Wednesday: 7,540,115 DKK
      • Thursday: 36,059,678 DKK
      • Friday: 11,823,366 DKK
      • Saturday: 10,593,359 DKK
      • Sunday: 6,346,358 DKK

⚠️ DATA GAPS FOR COST OPTIMIZATION:
--------------------------------------------------
   • No labor cost data (hourly wages)
   • No staff productivity metrics
   • No service time measurements
   • No customer satisfaction scores
   • No wait time tracking

🏪 6. VENUE-SPECIFIC DEMAND PATTERNS

📊 TOP 10 VENUES BY ORDER VOLUME:
--------------------------------------------------
   Kaffestuen Vesterbro           | 

In [11]:
# =============================================================================
# SOLUTION ARCHITECTURE SUMMARY
# =============================================================================
print("\n" + "=" * 80)
print("🏗️ SHIFT WIZARD: RECOMMENDED SOLUTION ARCHITECTURE")
print("=" * 80)

solution_components = """
┌─────────────────────────────────────────────────────────────────────────────┐
│                        🧙 SHIFT WIZARD ARCHITECTURE                          │
├─────────────────────────────────────────────────────────────────────────────┤
│                                                                             │
│  ┌─────────────────┐    ┌─────────────────┐    ┌─────────────────┐         │
│  │ 📊 DEMAND       │    │ 👥 STAFFING     │    │ 🚨 REAL-TIME    │         │
│  │ FORECASTER      │───▶│ OPTIMIZER       │───▶│ ADJUSTER        │         │
│  │                 │    │                 │    │                 │         │
│  │ • Historical    │    │ • Role matching │    │ • Call-off mgmt │         │
│  │   orders        │    │ • Capacity calc │    │ • SMS/Push      │         │
│  │ • Hour/Day/Week │    │ • Revenue-based │    │ • Replacement   │         │
│  │ • Venue-level   │    │   priority      │    │   finder        │         │
│  │ • Campaign aware│    │                 │    │                 │         │
│  └─────────────────┘    └─────────────────┘    └─────────────────┘         │
│           │                      │                      │                   │
│           ▼                      ▼                      ▼                   │
│  ┌───────────────────────────────────────────────────────────────┐         │
│  │                    📱 SHIFT WIZARD DASHBOARD                   │         │
│  │  • Weekly schedule view    • Coverage alerts                  │         │
│  │  • Demand heatmaps         • One-click adjustments            │         │
│  │  • Staff availability      • Cost vs Service metrics          │         │
│  └───────────────────────────────────────────────────────────────┘         │
│                                                                             │
└─────────────────────────────────────────────────────────────────────────────┘
"""
print(solution_components)

# Data availability summary
print("\n" + "=" * 80)
print("📋 DATA AVAILABILITY MATRIX")
print("=" * 80)

data_matrix = pd.DataFrame({
    'Feature': [
        'Demand Forecasting (Historical)',
        'Hour/Day Patterns',
        'Venue-Level Analysis',
        'Campaign Impact',
        'Staff Role Matching',
        'Contact Directory',
        'Venue Capacity',
        'Revenue Optimization',
        'Real-time Availability',
        'Labor Cost Analysis',
        'Employee Preferences',
        'External Events/Weather',
        'Service Time Metrics',
        'Call-off History'
    ],
    'Data Available': ['✅', '✅', '✅', '✅', '✅', '✅', '✅', '✅', '❌', '❌', '❌', '❌', '❌', '❌'],
    'Source Table': [
        'fct_orders',
        'fct_orders (created)',
        'fct_orders + dim_places',
        'fct_campaigns',
        'dim_users (roles)',
        'dim_users (phone/email)',
        'dim_tables (capacity)',
        'fct_orders (total_amount)',
        'Need new data',
        'Need new data',
        'Need new data',
        'External APIs',
        'Need new data',
        'Need new data'
    ]
})

display(data_matrix)

# What can be built NOW vs LATER
print("\n" + "=" * 80)
print("🎯 IMPLEMENTATION ROADMAP")
print("=" * 80)
print("""
┌─────────────────────────────────────────────────────────────────────────────┐
│ PHASE 1: BUILD NOW (Data Available)                                         │
├─────────────────────────────────────────────────────────────────────────────┤
│ ✅ Demand forecasting model (orders by hour/day/venue)                       │
│ ✅ Optimal staff calculator based on historical order patterns               │
│ ✅ Campaign-aware scheduling (surge prediction)                              │
│ ✅ Staff directory with role-based replacement matching                      │
│ ✅ Revenue-weighted shift prioritization                                     │
│ ✅ Venue-specific demand profiles                                            │
└─────────────────────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────────────────────────────────────────────┐
│ PHASE 2: COLLECT DATA FOR (Need New Data Sources)                           │
├─────────────────────────────────────────────────────────────────────────────┤
│ 📝 Staff availability preferences (new input system)                         │
│ 📝 Shift schedule history (track actual vs planned)                          │
│ 📝 Call-off tracking (patterns, reliability scores)                          │
│ 📝 Labor cost data (hourly rates by role)                                    │
│ 📝 Service time metrics (orders processed per staff per hour)                │
└─────────────────────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────────────────────────────────────────────┐
│ PHASE 3: EXTERNAL INTEGRATIONS (APIs & Feeds)                               │
├─────────────────────────────────────────────────────────────────────────────┤
│ 🌐 Weather API (OpenWeather, DMI) → Impact on customer traffic               │
│ 🌐 Local events calendar → Sports, concerts, holidays                        │
│ 🌐 Social media monitoring → Viral item detection                            │
│ 🌐 SMS/Push notification service → Rapid staff communication                 │
└─────────────────────────────────────────────────────────────────────────────┘
""")


🏗️ SHIFT WIZARD: RECOMMENDED SOLUTION ARCHITECTURE

┌─────────────────────────────────────────────────────────────────────────────┐
│                        🧙 SHIFT WIZARD ARCHITECTURE                          │
├─────────────────────────────────────────────────────────────────────────────┤
│                                                                             │
│  ┌─────────────────┐    ┌─────────────────┐    ┌─────────────────┐         │
│  │ 📊 DEMAND       │    │ 👥 STAFFING     │    │ 🚨 REAL-TIME    │         │
│  │ FORECASTER      │───▶│ OPTIMIZER       │───▶│ ADJUSTER        │         │
│  │                 │    │                 │    │                 │         │
│  │ • Historical    │    │ • Role matching │    │ • Call-off mgmt │         │
│  │   orders        │    │ • Capacity calc │    │ • SMS/Push      │         │
│  │ • Hour/Day/Week │    │ • Revenue-based │    │ • Replacement   │         │
│  │ • Venue-level   │    │   priority      │    │   finder        │         

Feature Data Available               Source Table
0   Demand Forecasting (Historical)              ✅                 fct_orders
1                 Hour/Day Patterns              ✅       fct_orders (created)
2              Venue-Level Analysis              ✅    fct_orders + dim_places
3                   Campaign Impact              ✅              fct_campaigns
4               Staff Role Matching              ✅          dim_users (roles)
5                 Contact Directory              ✅    dim_users (phone/email)
6                    Venue Capacity              ✅      dim_tables (capacity)
7              Revenue Optimization              ✅  fct_orders (total_amount)
8            Real-time Availability              ❌              Need new data
9               Labor Cost Analysis              ❌              Need new data
10             Employee Preferences              ❌              Need new data
11          External Events/Weather              ❌              External APIs
12             Service Time Metrics              ❌              Need new data
13                 Call-off History              ❌              Need new data


🎯 IMPLEMENTATION ROADMAP

┌─────────────────────────────────────────────────────────────────────────────┐
│ PHASE 1: BUILD NOW (Data Available)                                         │
├─────────────────────────────────────────────────────────────────────────────┤
│ ✅ Demand forecasting model (orders by hour/day/venue)                       │
│ ✅ Optimal staff calculator based on historical order patterns               │
│ ✅ Campaign-aware scheduling (surge prediction)                              │
│ ✅ Staff directory with role-based replacement matching                      │
│ ✅ Revenue-weighted shift prioritization                                     │
│ ✅ Venue-specific demand profiles                                            │
└─────────────────────────────────────────────────────────────────────────────┘

┌─────────────────────────────────────────────────────────────────────────────┐
│ PHASE 2: COLLECT DATA FOR (Need New Data Sources)                           │
├───────────

## 🧙 Shift Wizard Solution Summary

### What the Data Tells Us

| Insight | Value | Staffing Impact |
|---------|-------|-----------------|
| **Peak Hour** | 16:00 (4 PM) | Max staffing at dinner rush |
| **Busiest Day** | Friday (79K orders) | Double staff on Fridays |
| **Slowest Day** | Sunday (42K orders) | Reduce weekend shifts |
| **Order Mix** | 87% Takeaway, 13% Dine-in | Focus on kitchen/counter staff |
| **Top Venue** | Kaffestuen (25K orders) | Needs dedicated scheduler |
| **Staff Pool** | 1,887 operational staff | Sufficient for rotation |

### Solution Architecture (Based on Available Data)

```
┌──────────────────┐     ┌──────────────────┐     ┌──────────────────┐
│  DEMAND ENGINE   │────▶│ STAFF OPTIMIZER  │────▶│ ALERT SYSTEM     │
│                  │     │                  │     │                  │
│ • fct_orders     │     │ • dim_users      │     │ • Contact info   │
│ • Hour patterns  │     │ • roles column   │     │ • Role matching  │
│ • Day patterns   │     │ • dim_tables     │     │ • SMS/Email      │
│ • Campaigns      │     │ • Capacity       │     │                  │
└──────────────────┘     └──────────────────┘     └──────────────────┘
```

### ✅ What We CAN Build Now
1. **Demand Forecasting Model** - Using 400K historical orders
2. **Hour/Day Heatmaps** - Peak detection at venue level
3. **Campaign Surge Planning** - 641 campaigns with day-specific timing
4. **Staff Role Matching** - 1,771 cashiers, 63 managers available
5. **Revenue-Weighted Scheduling** - Prioritize high-value time slots

### ❌ Data Gaps (Need to Collect)
1. **Staff Preferences** - Availability, preferred shifts
2. **Labor Costs** - Hourly rates, overtime rules
3. **Call-off History** - No tracking of absences
4. **Service Metrics** - Orders per staff per hour
5. **External Factors** - Weather, events, social media

### Recommended MVP Features
1. 📊 **Demand Dashboard** - Hourly/daily order predictions per venue
2. 👥 **Smart Scheduler** - Auto-suggest staff based on predicted demand
3. 🚨 **Gap Alerts** - Notify when coverage drops below threshold
4. 📱 **Quick Replace** - One-tap replacement finder with contact info